fastjet/internal/delaunay/triangle.go

```diff
@@ -0,0 +1,147 @@
+func (t *Triangle) add() (updated []*Point) {
+	t.isInTriangulation = true
+	// update the adjacent lists
+	t.A.adjacentTriangles = append(t.A.adjacentTriangles, t)
+	t.B.adjacentTriangles = append(t.B.adjacentTriangles, t)
+	t.C.adjacentTriangles = append(t.C.adjacentTriangles, t)
+	// update nearest neighbor if one of the points is closer than current nearest.
+	// First find the nearest neighbor in t.
+	// Then check whether local min distance is smaller than current min distance.
+	distAB := t.A.distance(t.B)
+	distBC := t.B.distance(t.C)
+	distCA := t.C.distance(t.A)
+	var localMin float64
+	var localNearest *Point
+	if distAB < distCA {
+		localMin = distAB
+		localNearest = t.B
+	} else {
+		localMin = distCA
+		localNearest = t.C
+	}
+	if localMin < t.A.dist {
+		t.A.dist = localMin
+		t.A.nearest = localNearest
+		updated = append(updated, t.A)
+	}
+	if distAB < distBC {
+		localMin = distAB
+		localNearest = t.A
+	} else {
+		localMin = distBC
+		localNearest = t.C
+	}
+	if localMin < t.B.dist {
+		t.B.dist = localMin
+		t.B.nearest = localNearest
+		updated = append(updated, t.B)
+	}
+	if distBC < distCA {
+		localMin = distBC
+		localNearest = t.B
+	} else {
+		localMin = distCA
+		localNearest = t.A
+	}
+	if localMin < t.C.dist {
+		t.C.dist = localMin
+		t.C.nearest = localNearest
+		updated = append(updated, t.C)
+	}
+	return updated
+}
+
+// remove is used when the triangle t is remove from the delaunay triangulation.
+// It updates the information of the points in t.
+// It returns all points who's nearest neighbor was updated.
+func (t *Triangle) remove() (updated []*Point) {
+	t.isInTriangulation = false
+	t.A.adjacentTriangles = t.A.adjacentTriangles.remove(t)
+	t.B.adjacentTriangles = t.B.adjacentTriangles.remove(t)
+	t.C.adjacentTriangles = t.C.adjacentTriangles.remove(t)
+	// update the nearest neighbor if the nearest neighbor is in t.
+	if t.A.nearest.Equals(t.B) || t.A.nearest.Equals(t.C) {
+		t.A.findNearest()
+		updated = append(updated, t.A)
+	}
+	if t.B.nearest.Equals(t.A) || t.B.nearest.Equals(t.C) {
+		t.B.findNearest()
+		updated = append(updated, t.B)
+	}
+	if t.C.nearest.Equals(t.A) || t.C.nearest.Equals(t.B) {
+		t.C.findNearest()
+		updated = append(updated, t.C)
+	}
+	return updated
+}
+
+// Equals checks whether two triangles are the same.
+func (t *Triangle) Equals(s *Triangle) bool {
+	return t == s ||
+		(t.A.Equals(s.A) || t.A.Equals(s.B) || t.A.Equals(s.C)) &&
+			(t.B.Equals(s.A) || t.B.Equals(s.B) || t.B.Equals(s.C)) &&
+			(t.C.Equals(s.A) || t.C.Equals(s.B) || t.C.Equals(s.C))
+}
+
+func (t *Triangle) String() string {
+	return fmt.Sprintf("{A%s, B%s, C%s}", t.A, t.B, t.C)
+}
+
+type triangles []*Triangle
+
+// remove removes given triangles from a slice of triangles
+func (t triangles) remove(triangles ...*Triangle) triangles {
+	for i := len(t) - 1; i >= 0; i-- {
+		for j, triangleToRemove := range triangles {
+			if triangleToRemove.Equals(t[i]) {
+				t = append(t[:i], t[i+1:]...)
```

Comment:
do we need to keep the order of the already iterated elements of `t` ?
otherwise we could just use:
```go
n := len(t)-1
t[i], t[n] = t[n], nil
t = t[:n]
```